gems/smithy-client/lib/smithy-client/http/response.rb

```diff
@@ -1,39 +1,168 @@
+        # @return [Headers]
+        attr_accessor :headers
+
+        # @return [IO, StringIO]
+        attr_reader :body
+
+        # @return [StandardError, nil]
+        attr_reader :error
+
+        # @param [#read, #size, #rewind] io
+        def body=(io)
+          @body =
+            case io
+            when nil then StringIO.new('')
+            when String then StringIO.new(io)
+            else io
+            end
+        end
+
+        # @param [Integer] status_code
+        # @param [Hash<String, String>] headers
+        def signal_headers(status_code, headers)
+          @status_code = status_code
+          @headers = Headers.new(headers)
+          emit(:headers, @status_code, @headers)
+        end
+
+        # @param [string] chunk
+        def signal_data(chunk)
+          return if chunk == ''
+
+          @body.write(chunk)
+          emit(:data, chunk)
+        end
+
+        # Completes the http response.
+        #
+        # @example Completing the response in a single call
+        #
+        #     http_response.signal_done(
+        #       status_code: 200,
+        #       headers: {},
+        #       body: ''
+        #     )
+        #
+        # @example Complete the response in parts
+        #
+        #     # signal headers straight-way
+        #     http_response.signal_headers(200, {})
+        #
+        #     # signal data as it is received from the socket
+        #     http_response.signal_data("...")
+        #     http_response.signal_data("...")
+        #     http_response.signal_data("...")
+        #
+        #     # signal done once the body data is all written
+        #     http_response.signal_done
+        #
+        # @overload signal_done()
+        #
+        # @overload signal_done(options = {})
+        #   @option options [required, Integer] :status_code
+        #   @option options [required, Hash] :headers
+        #   @option options [required, String] :body
+        #
+        def signal_done(options = {})
+          if options.keys.sort == %i[body headers status_code]
+            signal_headers(options[:status_code], options[:headers])
+            signal_data(options[:body])
+            signal_done
+          elsif options.empty?
+            @body.rewind if @body.respond_to?(:rewind)
+            @done = true
+            emit(:done)
+          else
+            msg = 'options must be empty or must contain :status_code, :headers, ' \
+                  'and :body'
+            raise ArgumentError, msg
+          end
+        end
+
+        # @param [StandardError] error
+        def signal_error(error)
+          @error = error
+          signal_done
+        end
+
+        def on_headers(status_code_range = nil, &block)
+          @listeners[:headers] << listener(status_code_range, block)
+        end
+
+        def on_data(&callback)
+          @listeners[:data] << callback
+        end
+
+        def on_done(status_code_range = nil, &callback)
```

Comment:
I know this is what we do in V3 - but do we ever use the status code range in events?  To me this seems to add a lot of extra complexity that I'd rather avoid unless we think we really need it.  I already would prefer to not have events since they add yet another path for injecting logic into the lifecycle, but if we are adding them, we should at least try and use the simplest interface.

Reply:
Yes, we do use most of them. On headers we look at different ranges for when we do streaming retry. S3 responds with errors in 200 range so sometimes that range is necessary. We currently have listeners for all three signals (headers, data, done) and wither success (data) or errors (error) with ranges to handle weirdness. I think it makes sense to have them at these levels.